packages/lesswrong/server/collections/bookmarks/mutations.ts

```diff
@@ -0,0 +1,100 @@
+import gql from 'graphql-tag';
+import { Bookmarks } from './collection';
+import { userCanDo } from "@/lib/vulcan-users/permissions";
+import { logFieldChanges } from "@/server/fieldChanges";
+
+export const bookmarkGqlTypeDefs = gql`
+  input ToggleBookmarkInput {
+    documentId: String!
+    collectionName: String!
+  }
+
+  type ToggleBookmarkOutput {
+    data: Bookmark
+  }
+
+  extend type Mutation {
+    toggleBookmark(input: ToggleBookmarkInput!): ToggleBookmarkOutput
+  }
+`;
+
+interface ToggleBookmarkInput {
+  documentId: string;
+  collectionName: "Posts" | "Comments";
+}
+
+async function toggleBookmarkResolver(root: void, { input }: { input: ToggleBookmarkInput }, context: ResolverContext): Promise<{ data: DbBookmark | null }> {
+  const { documentId, collectionName } = input;
+  const { currentUser } = context;
+
+  if (!currentUser || !userCanDo(currentUser, 'bookmarks.toggle')) {
+    throw new Error("Permission denied. You must be logged in to bookmark items.");
+  }
+
+  if (!documentId || !collectionName || !["Posts", "Comments"].includes(collectionName)) {
+    throw new Error("Invalid input: documentId and collectionName (Posts or Comments) are required.");
+  }
+
+  const existingBookmark = await Bookmarks.findOne({
```

Comment:
This entire operation should be a single database-level upsert (see `upsertReadStatus`), which requires a unique index on `userId, documentId, collectionName`.  otherwise you get race conditions.